pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PublishRateLimiterImpl.java

```diff
@@ -16,70 +16,133 @@
+        if (currentTokenBucketOnByte != null) {
+            // consume tokens from the token bucket for bytes
+            currentTokenBucketOnByte.consumeTokens(msgSizeInBytes);
+            // check if the token bucket contains remaining tokens, if not, we should throttle
+            shouldThrottle = shouldThrottle || !currentTokenBucketOnByte.containsTokens();
+        }
+        if (shouldThrottle) {
+            // throttle the producer by incrementing the throttle count
+            producer.incrementThrottleCount();
+            // schedule decrementing the throttle count to possibly unthrottle the producer after the
+            // throttling period
+            scheduleDecrementThrottleCount(producer);
         }
     }
 
-    @Override
-    public void incrementPublishCount(int numOfMessages, long msgSizeInBytes) {
-        if (this.publishThrottlingEnabled) {
-            this.currentPublishMsgCount.add(numOfMessages);
-            this.currentPublishByteCount.add(msgSizeInBytes);
+    private void scheduleDecrementThrottleCount(Producer producer) {
+        // add the producer to the queue of producers to be unthrottled
+        unthrottlingQueue.offer(producer);
+        // schedule unthrottling if not already scheduled
+        if (unthrottlingScheduled.compareAndSet(false, true)) {
+            EventLoopGroup executor = producer.getCnx().getBrokerService().executor();
+            scheduleUnthrottling(executor);
         }
     }
 
-    @Override
-    public boolean resetPublishCount() {
-        if (this.publishThrottlingEnabled) {
-            this.currentPublishMsgCount.reset();
-            this.currentPublishByteCount.reset();
-            this.publishRateExceeded = false;
-            return true;
+    private void scheduleUnthrottling(ScheduledExecutorService executor) {
+        executor.schedule(() -> this.unthrottleQueuedProducers(executor), calculateThrottlingDurationNanos(),
+                TimeUnit.NANOSECONDS);
+    }
+
+    private long calculateThrottlingDurationNanos() {
+        AsyncTokenBucket currentTokenBucketOnMessage = tokenBucketOnMessage;
+        long throttlingDurationNanos = 0L;
+        if (currentTokenBucketOnMessage != null) {
+            throttlingDurationNanos = currentTokenBucketOnMessage.calculateThrottlingDuration();
+        }
+        AsyncTokenBucket currentTokenBucketOnByte = tokenBucketOnByte;
+        if (currentTokenBucketOnByte != null) {
+            throttlingDurationNanos = Math.max(throttlingDurationNanos,
+                    currentTokenBucketOnByte.calculateThrottlingDuration());
         }
-        return false;
+        return throttlingDurationNanos;
     }
 
-    @Override
-    public boolean isPublishRateExceeded() {
-        return publishRateExceeded;
+    private void unthrottleQueuedProducers(ScheduledExecutorService executor) {
+        Producer producer;
+        // unthrottle producers until the token buckets contain tokens
```

Comment:
**[Nit]** I think this should be  "unthrottle producers until the token buckets do not contain tokens" or "unthrottle as many producers as possible while there are token available".

Reply:
Thanks. Applied.